File: server/routes/bioskop.js
import express from 'express';
import pool from '../config/database.js';

const router = express.Router();

// Get all bioskops
router.get('/', async (req, res) => {
  try {
    const result = await pool.query('SELECT * FROM bioskops ORDER BY id');
    res.json(result.rows);
  } catch (error) {
    console.error('Error fetching bioskops:', error);
    res.status(500).json({ error: 'Failed to fetch bioskops' });
  }
});

// Get bioskop by ID
router.get('/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const result = await pool.query('SELECT * FROM bioskops WHERE id = $1', [id]);
    
    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Bioskop not found' });
    }
    
    res.json(result.rows[0]);
  } catch (error) {
    console.error('Error fetching bioskop:', error);
    res.status(500).json({ error: 'Failed to fetch bioskop' });
  }
});

// Create new bioskop
router.post('/', async (req, res) => {
  try {
    const { nama, lokasi, rating } = req.body;
    
    if (!nama || !lokasi) {
      return res.status(400).json({ error: 'Nama dan Lokasi tidak boleh kosong' });
    }
    
    const result = await pool.query(
      'INSERT INTO bioskops (nama, lokasi, rating) VALUES ($1, $2, $3) RETURNING *',
      [nama, lokasi, rating || 4.5]
    );
    
    res.status(201).json(result.rows[0]);
  } catch (error) {
    console.error('Error creating bioskop:', error);
    res.status(500).json({ error: 'Failed to create bioskop' });
  }
});

// Update bioskop
router.put('/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const { nama, lokasi, rating } = req.body;
    
    const result = await pool.query(
      'UPDATE bioskops SET nama = $1, lokasi = $2, rating = $3, updated_at = CURRENT_TIMESTAMP WHERE id = $4 RETURNING *',
      [nama, lokasi, rating, id]
    );
    
    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Bioskop not found' });
    }
    
    res.json(result.rows[0]);
  } catch (error) {
    console.error('Error updating bioskop:', error);
    res.status(500).json({ error: 'Failed to update bioskop' });
  }
});

// Delete bioskop
router.delete('/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const result = await pool.query('DELETE FROM bioskops WHERE id = $1 RETURNING *', [id]);
    
    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Bioskop not found' });
    }
    
    res.json({ message: 'Bioskop deleted successfully' });
  } catch (error) {
    console.error('Error deleting bioskop:', error);
    res.status(500).json({ error: 'Failed to delete bioskop' });
  }
});

export default router;

File: controllers/bioskop.go
package controllers

import (
	"bioskop-app/config"
	"bioskop-app/models"
	"net/http"

	"github.com/gin-gonic/gin"
)

func CreateBioskop(c *gin.Context) {
	var bioskop models.Bioskop

	// Bind dan validasi input
	if err := c.ShouldBindJSON(&bioskop); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Nama dan Lokasi tidak boleh kosong"})
		return
	}

	// Simpan ke DB
	if err := config.DB.Create(&bioskop).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Gagal menyimpan bioskop"})
		return
	}

	c.JSON(http.StatusCreated, bioskop)
}

func GetBioskops(c *gin.Context) {
	var bioskops []models.Bioskop

	if err := config.DB.Find(&bioskops).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch bioskops"})
		return
	}

	c.JSON(http.StatusOK, bioskops)
}

func GetBioskop(c *gin.Context) {
	id := c.Param("id")
	var bioskop models.Bioskop

	if err := config.DB.First(&bioskop, id).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Bioskop not found"})
		return
	}

	c.JSON(http.StatusOK, bioskop)
}

func UpdateBioskop(c *gin.Context) {
	id := c.Param("id")
	var bioskop models.Bioskop

	if err := config.DB.First(&bioskop, id).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Bioskop not found"})
		return
	}

	if err := c.ShouldBindJSON(&bioskop); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	if err := config.DB.Save(&bioskop).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update bioskop"})
		return
	}

	c.JSON(http.StatusOK, bioskop)
}

func DeleteBioskop(c *gin.Context) {
	id := c.Param("id")
	
	if err := config.DB.Delete(&models.Bioskop{}, id).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete bioskop"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Bioskop deleted successfully"})
}